src/django_pg_migration_tools/operations.py

```diff
@@ -226,5 +236,185 @@ def database_backwards(
             to_state=to_state,
             index=index,
             model=model,
-            operation=self,
+            unique=False,
+        )
+
+
+class ConstraintOperationError(Exception):
+    pass
+
+
+class ConstraintAlreadyExists(ConstraintOperationError):
+    pass
+
+
+class SaferAddUniqueConstraint(BaseIndexOperation, operation_models.AddConstraint):
```

Comment:
This example code is very rough, but I feel it would be clearer what is part of the classes you're inheriting and what is part of the logic you're adding on top if that logic was isolated

<details>
<summary>suggestion</summary>

```
class _SaferAddUniqueConstraintLogic:
    model_name: str
    constraint: models.UniqueConstraint
    raise_if_exists: bool
    index_name: str

    CHECK_EXISTING_CONSTRAINT_QUERY: ClassVar[str] = dedent("""
        SELECT conname
        FROM pg_catalog.pg_constraint
        WHERE conname = %(constraint_name)s;
    """)

    def __init__(
        self,
        model_name: str,
        constraint: models.UniqueConstraint,
        raise_if_exists: bool = True,
    ) -> None:
        self.raise_if_exists = raise_if_exists
        self.constraint = constraint
        self._index_name = f"{constraint.name}_idx"

    def execute_forwards(
        self,
        *,
        model: object,
        schema_editor: base_schema.BaseDatabaseSchemaEditor,
        index: str,
    ) -> None:
        # Django doesn't have a handy flag "using=..." so we need to alter the
        # SQL statement manually. We go from a SQL that looks like this:
        #
        # - ALTER TABLE "table" ADD CONSTRAINT "constraint" UNIQUE ("field")
        #
        # Into a SQL that looks like:
        #
        # - ALTER TABLE "table" ADD CONSTRAINT "constraint" UNIQUE USING INDEX "idx"
        base_sql = str(self.constraint.create_sql(model, schema_editor))
        alter_table_sql = base_sql.split(" UNIQUE")[0]
        sql = f'{alter_table_sql} UNIQUE USING INDEX "{index.name}"'

        # Run the sql to add the constraint inside a transaction. We have lock
        # timeouts back in place after creating the index that would prevent
        # this operation from running for too long if it's blocked by another
        # query. Otherwise, this operation should actually be quite fast - if
        # it's not blocked - since we have created the unique index in the
        # previous step.
        with transaction.atomic():
            return schema_editor.execute(sql)

    def constraint_exists(
        self,
        schema_editor: base_schema.BaseDatabaseSchemaEditor,
    ) -> bool:
        cursor = schema_editor.connection.cursor()
        cursor.execute(
            self._CHECK_EXISTING_CONSTRAINT_QUERY,
            {"constraint_name": self.constraint.name},
        )
        return bool(cursor.fetchone())

    def can_create_constraint(
        self,
        schema_editor: base_schema.BaseDatabaseSchemaEditor,
    ) -> bool:
        constraint_exists = self._constraint_exists(schema_editor)
        if self.raise_if_exists and constraint_exists:
            raise ConstraintAlreadyExists(
                f"Cannot create a constraint with the name "
                f"{self.constraint.name} because a constraint of the same "
                f"name already exists. If you want to skip this operation "
                f"when the constraint already exists, run the operation "
                f"with the flag `skip_if_exists=True`."
            )
        # We can't re-create a constraint that already exists because the
        # ALTER TABLE ... ADD CONSTRAINT is not idempotent.
        return not constraint_exists

    def get_index_for_constraint(self) -> models.Index:
        return models.Index(
            *self.constraint.expressions,
            fields=self.constraint.fields,
            name=self._index_name,
            condition=self.constraint.condition,
            opclasses=self.constraint.opclasses,  # type: ignore[attr-defined]
        )

    def validate(self) -> None:
        if not isinstance(self.constraint, models.UniqueConstraint):
            raise ValueError(
                "SaferAddUniqueConstraint only supports the UniqueConstraint class"
            )

        if len(self._index_name) > MAX_RELATION_NAME_LEN:
            raise ValueError(
                f"Index name {self._index_name} is longer than "
                f"{MAX_RELATION_NAME_LEN} characters limit."
            )


class SaferAddUniqueConstraint(BaseIndexOperation, operation_models.AddConstraint):
    logic: _SaferAddUniqueConstraintLogic

    def __init__(
        self,
        model_name: str,
        constraint: models.UniqueConstraint,
        raise_if_exists: bool = True,
    ) -> None:
        self._logic = _SaferAddUniqueConstraintLogic(
            model_name=model_name,
            constraint=constraint,
            raise_if_exists=raise_if_exists,
        )

        # Perform a basic input-validation at initialisation time rather than
        # at migration time ("database_forwards"/"database_backwards") so that
        # all the errors we __can__ raise at initialisation time are raised
        # when the programmer is still coding rather than when the migration
        # effectively runs. This is the "Django way" of doing it, as other
        # operation classes take this same approach. See "AddIndex" operation.
        self._logic.validate()

        super().__init__(model_name, constraint)

    def database_forwards(
        self,
        app_label: str,
        schema_editor: base_schema.BaseDatabaseSchemaEditor,
        from_state: migrations.state.ProjectState,
        to_state: migrations.state.ProjectState,
    ) -> None:
        if not self._can_migrate(app_label, schema_editor, from_state):
            return

        if not self._logic.can_create_constraint(schema_editor):
            return

        index = self._logic.get_index_for_constraint()
        model = to_state.apps.get_model(app_label, self.model_name)
        self.safer_create_index(
            app_label=app_label,
            schema_editor=schema_editor,
            from_state=from_state,
            to_state=to_state,
            index=index,
            model=model,
            unique=True,
        )

        self._logic.execute_forwards(
            model=model, schema_editor=schema_editor, index=index
        )

    def database_backwards(
        self,
        app_label: str,
        schema_editor: base_schema.BaseDatabaseSchemaEditor,
        from_state: migrations.state.ProjectState,
        to_state: migrations.state.ProjectState,
    ) -> None:
        if not self._can_migrate(app_label, schema_editor, from_state):
            return

        if not self._logic.constraint_exists(schema_editor):
            # Nothing to delete.
            return

        # Drop the constraint in a transaction so that it can raise a lock
        # timeout error if it takes too long. This transaction.atomic ctx
        # manager is required because the "forward" part of the migration
        # requires an atomic=False flag.
        with transaction.atomic():
            super().database_backwards(
                app_label=app_label,
                schema_editor=schema_editor,
                from_state=from_state,
                to_state=to_state,
            )

    def _can_migrate(
        self,
        app_label: str,
        schema_editor: base_schema.BaseDatabaseSchemaEditor,
        from_state: migrations.state.ProjectState,
    ) -> bool:
        model = from_state.apps.get_model(app_label, self.model_name)
        return bool(self.allow_migrate_model(schema_editor.connection.alias, model))

    def describe(self) -> str:
        return (
            f"Concurrently adds a UNIQUE index {self._index_name} on model "
            f"{self.model_name} on field(s) {self.constraint.fields} if the "
            f"index does not exist. Then, adds the constraint using the just-created "
            f"index. NOTE: Using django_pg_migration_tools SaferAddUniqueConstraint "
            f"operation."
        )

```
</details>

Reply:
@delfick How about using composition over of inheritance? Making so that the `BaseIndexOperation` is the part being composed instead?


```python
class SaferAddIndexConcurrently(psql_operations.AddIndexConcurrently):
    def __init__(self, args, kwargs):
        self.index_operator = BaseIndexOperation()
        ....

class SaferRemoveIndexConcurrently(psql_operations.AddIndexConcurrently):
    def __init__(self, args, kwargs):
        self.index_operator = BaseIndexOperation()
        ....

class SaferAddUniqueConstraint(...):
    def __init__(self, args, kwargs):
        self.index_operator = BaseIndexOperation()
        ....
```

Alternatively, we can just turn those into class methods and use that class directly:

```python
class SaferRemoveIndexConcurrently(psql_operations.AddIndexConcurrently):
    def database_forwards(...):
        BaseIndexOperation.safer_create_index(...)
```

That achieves what you are suggesting by making it explicit when we are using "outside logic", and we get away from multiple inheritance.

Reply:
yeah, exactly, the point is to have less things combined in one place.

Reply:
Thanks, I've created this issue to track that work -> https://github.com/kraken-tech/django-pg-migration-tools/issues/28

Reply:
ah yeah, with the intention of merging this and doing that kind of transformation on this class as separate PR after this one?

Reply:
Yep 👍🏽 reminding that the refactor will touch more than this class as it needs to bring the other two operations inheriting from the BaseIndexOperation along. I also want to rename a few things as that base class doesn't feel like the right name so will do that in a follow up.